assignments/burmanm/week-4.scm

```diff
@@ -0,0 +1,208 @@
+	(else (cons numer denumer))))
+
+;; Exercise 2.2
+
+(define (make-segment p1 p2)
+  (cons p1 p2))
+
+(define (start-segment segment)
+  (car segment))
+
+(define (end-segment segment)
+  (cdr segment))
+
+(define (make-point x y)
+  (cons x y))
+
+(define (x-point p)
+  (car p))
+
+(define (y-point p)
+  (cdr p))
+
+(define (midpoint-segment segment)
+  (make-point
+   (/ (+ (x-point (start-segment segment)) (x-point (end-segment segment))) 2)
+   (/ (+ (y-point (start-segment segment)) (y-point (end-segment segment))) 2)))
+
+(define (print-point p)
+  (newline)
+  (display "(")
+  (display (x-point p))
+  (display ",")
+  (display (y-point p))
+  (display ")"))
+
+;; Test:
+
+(print-point
+ (midpoint-segment (make-segment (make-point 0 0) (make-point 4 4))))
+
+;; (2,2)
+
+;; Exercise 2.3
+
+;; Use segments above, two opposite segments
+(define (make-rectangle seg1 seg2)
+  (cons seg1 seg2))
+
+(define (calc-lengths rec)
+  (cons
+   (- (x-point (start-segment (car rec))) (x-point (start-segment (cdr rec))))
+   (- (y-point (start-segment (car rec))) (y-point (end-segment (car rec))))))
+
+(define (perimeter-rectangle rec)
+  (let* ((lengths (calc-lengths rec))
+	 (x (car lengths))
+	 (y (cdr lengths))
+	 (calc-rectangle
+	  (* 2 (+ x y))))
+  (if (< calc-rectangle 0)
+      (* -1 calc-rectangle)
+      calc-rectangle)))
+
+;; Test
+(perimeter-rectangle
+ (make-rectangle (make-segment (make-point 0 0) (make-point 0 4)) (make-segment (make-point 10 0) (make-point 10 4))))
+
+;; $21 = 28
+
+(define (area-rectangle rec)
+  (let* ((lengths (calc-lengths rec))
+	 (x (car lengths))
+	 (y (cdr lengths))
+	 (calc-area (* x y)))
+    calc-area))
+
+;; Test
+(area-rectangle
+ (make-rectangle (make-segment (make-point 0 0) (make-point 0 4)) (make-segment (make-point 10 0) (make-point 10 4))))
+
+;; $23 = 40
+
+;; Exercise 2.4
+
+(define (cdr z)
+  (z (lambda (p q) q)))
+
+;; (define (cons x y)
+;;   (lambda (m) (m x y)))
+;; (define (car z)
+;;   (z (lambda (p q) p)))
+;; (car (cons x y))
+;;
+;; (car ((lambda (m) (m x y) x y)))
+;; ((z (lambda (p q) p)) ((lambda (m) (m x y) x y)))
+;; ((lambda (m) (m x y) x y) (lambda (p q) p))
+;; m is substituted with the (lambda (p q) p) which returns first argument of (x y)
+
+;; Exercise 2.5
+
+;; Math question again.. meh. 
+
+;; Exercise 2.6
+
+;; (define zero (lambda (f) (lambda (x) x)))
+;; (define (add-1 n)
+;;   (lambda (f) (lambda (x) (f ((n f) x)))))
+
+
+;; (add-1 zero)
+
+(define one (lambda (f) (lambda (x) (f ((lambda (x) x) x)))))
+
+;; call the last one, outer lambda with x returns x..
+(define one (lambda (f) (lambda (x) (f x))))
+
+;; (add-1 one) same idea
+(define two (lambda (f) (lambda (x) (f (f x)))))
+
+;; +
+
+;; Is basically calling add-1 n-times. That is, (f ((n f) x)) * n-times. Since n is used, use m:
+
+(define + (x y) (lambda (f) (lambda (x) ((m f) ((n f) x)))))
+
+;; Exercise 2.7
```

Comment:
Note that the initial schedule didn't include section 2.4 for this 4th week. I'll do mine shortly.